File: src/controls/select.ts
import { observable, register } from "../core/decorators";
import { Control } from "./control";

// Wraps a `<select>`. Options present in the markup at adopt time are
// preserved as a prefix; programmatic `options` append after them.
@register
export class Select extends Control {
  @observable accessor options: string[] = [];
  @observable accessor value: string = "";

  private templateOptionCount = 0;
  private onChange = (): void => { this.value = this.selectEl.value; };

  constructor(el: HTMLElement) {
    super(el);
    this.templateOptionCount = this.selectEl.options.length;
    this.selectEl.addEventListener("change", this.onChange);
  }

  optionsDidChange(_prev: string[], next: string[]): void {
    const sel = this.selectEl;
    while (sel.options.length > this.templateOptionCount) {
      sel.remove(this.templateOptionCount);
    }
    for (const v of next) {
      const opt = document.createElement("option");
      opt.value = v;
      opt.textContent = v;
      sel.appendChild(opt);
    }
    if (this.value && Array.from(sel.options).some((o) => o.value === this.value)) {
      sel.value = this.value;
    }
  }

  valueDidChange(_prev: string, next: string): void {
    if (this.selectEl.value !== next) this.selectEl.value = next;
  }

  private get selectEl(): HTMLSelectElement {
    return this.element as HTMLSelectElement;
  }
}

File: opal/lib/swill/select.rb
# backtick_javascript: true

module Swill
  # Observable wrapper for a native <select>. Options already present in the
  # markup are retained as a prefix; assigning +options+ replaces only the
  # programmatic suffix, matching the TypeScript control.
  class Select < Control
    include Observable

    property :options, default: -> { [] }
    property :value, default: ""

    def initialize(element)
      super
      @template_option_count = `#{element}.children.length`
      initial = `#{element}.value`
      initial = `#{element}.getAttribute("value")` if (initial.nil? || initial == "") && `#{element}.hasAttribute("value")`
      self.value = initial.to_s
      @change_listener = lambda { |_event| self.value = `#{element}.value`.to_s }
      `#{element}.addEventListener("change", #{@change_listener})`
    end

    def options_did_change(_previous, values)
      while `#{element}.children.length > #{@template_option_count}`
        `#{element}.children[#{@template_option_count}].remove()`
      end

      values.each do |value|
        option = `document.createElement("option")`
        `#{option}.value = #{value.to_s}`
        `#{option}.textContent = #{value.to_s}`
        `#{element}.appendChild(#{option})`
      end
      sync_value
    end

    def value_did_change(_previous, _value)
      sync_value
    end

    def teardown!
      `#{element}.removeEventListener("change", #{@change_listener})` if @change_listener
    end

    private

    def sync_value
      `#{element}.value = #{value.to_s}` if `#{element}.value !== #{value.to_s}`
    end
  end
end
